compiler/rustc_mir_transform/src/upvar_to_local_prop.rs

```diff
@@ -0,0 +1,694 @@
+//! This pass propagates upvars to locals within a generator, if it can
+//! determine such coalescing is sound.
+//!
+//! Original demo: search for: `_3 = _1.0`; if found, then replace all occurrences of `_3` with `_1.0`.
+//!
+//! The generalization:
+//!
+//! * apply to more upvars than just `_1.0`; e.g. `_1.1`, `_1.2`, ... (but don't need to worry
+//!   about array indices nor enum variants).
+//!
+//! * transitively apply the replacement to subsequent copies of the local(s).
+//!   e.g. `_3 = _1.0; ... _4 = _3; ...; _5 = _4;`, then might replace `_5` with `_1.0`.
```

Comment:
This looks a lot like what `CopyProp` does. (That pass is very poorly named, as it merges locals more than anything.)
There may be a way to simplify the implementation by using the same analysis to unify locals, and a second pass to replace the locals by the upvars.

Reply:
Yes, its possible there's opportunity for simplification there.

Part of my goal here was to have this be an inherently *cheap* pass. (For example, I wanted to avoid any question of whether there might be a fixed point iteration going on.)

But if `CopyProp` is sufficiently cheap to run that we might consider turning it on for `mir-opt-level > 0`, I'd be willing to consider unifying these things.

Reply:
(I also was hoping for this to be an "obviously correct" pass, though the fact that I encountered a number of issues in its development has led me to think that maybe there's no such thing as "obviously correct" when it comes to MIR transformations... 🤔 )

Reply:
FYI `CopyProp` is already so enabled.

It sounds like you are realizing that MIR is not a good IR to write optimizations for, which is a very common observation.

Reply:
> It sounds like you are realizing that MIR is not a good IR to write optimizations for, which is a very common observation.

Well... I've seen much much worse. :)

Reply:
> But if CopyProp is sufficiently cheap to run that we might consider turning it on for mir-opt-level > 0, I'd be willing to consider unifying these things.

I've added investigation of this option as a to-do for the PR. (I.e., I'm going to attempt to revise `upvar_to_local_prop` so that it leverages CopyProp+SSA code in some way. Either it works and this PR gets vastly simpler, or it doesn't work and I provide an explanation for why to continue using the code as written here.)